qiskit/backends/_sympy_unitarysimulator.py

```diff
@@ -0,0 +1,280 @@
+        left_up = cos(theta/2)
+        right_up = (-E**(I*lamb)) * sin(theta/2)
+        left_down = (E**(I*phi)) * sin(theta/2)
+        right_down = (E**(I*(phi+lamb))) * cos(theta/2)
+        u_mat = Matrix([[left_up, right_up], [left_down, right_down]])
+
+        if theta_is_regular and phi_is_regular and lamb_is_regular:
+            u_mat_numeric = u_mat
+        else:
+            u_mat_numeric = u_mat.evalf()
+        return u_mat_numeric
+
+    @staticmethod
+    def compute_ugate_matrix_wrap(parafloatlist):
+        """preparation needed before computing the ugate matrix
+            Args:
+                parafloatlist (list): parafloatlist
+            Returns:
+                Matrix: matrix that represents the ugate
+        """
+        if len(parafloatlist) == 1:  # [theta=0, phi=0, lambda]
+            parafloatlist.insert(0, 0.0)
+            parafloatlist.insert(0, 0.0)
+        elif len(parafloatlist) == 2:  # [theta=pi/2, phi, lambda]
+            parafloatlist.insert(0, pi/2)
+        elif len(parafloatlist) == 3:  # [theta, phi, lambda]
+            pass
+        else:
+            return NotImplemented
+
+        u_mat = SympyUnitarySimulator.compute_ugate_matrix(parafloatlist)
+        return u_mat
+
+    def enlarge_single_opt_sympy(self, opt, qubit, number_of_qubits):
+        """Enlarge single operator to n qubits.
+
+        It is exponential in the number of qubits.
+
+        Args:
+            opt (object): the single-qubit opt.
+            qubit (int): the qubit to apply it on counts from 0 and order
+                is q_{n-1} ... otimes q_1 otimes q_0.
+            number_of_qubits (int): the number of qubits in the system.
+        Returns:
+            object: the enlarged tensor product.
+        """
+        temp_1 = eye(2**(number_of_qubits-qubit-1))
+        temp_2 = eye(2**(qubit))
+        enlarge_opt = TensorProduct(temp_1, TensorProduct(opt, temp_2))
+        return enlarge_opt
+
+    def _add_unitary_single(self, gate, qubit):
+        """Apply the single-qubit gate.
+
+        gate is the single-qubit gate.
+        qubit is the qubit to apply it on counts from 0 and order
+            is q_{n-1} ... otimes q_1 otimes q_0.
+        number_of_qubits is the number of qubits in the system.
+        """
+        unitaty_add = self.enlarge_single_opt_sympy(gate, qubit, self._number_of_qubits)
+        self._unitary_state = unitaty_add*self._unitary_state  # * means "dot product"
+
+    def enlarge_two_opt_sympy(self, opt, qubit0, qubit1, num):
+        """Enlarge two-qubit operator to n qubits.
+        It is exponential in the number of qubits.
+        opt is the two-qubit gate
+        q0 is the first qubit (control) counts from 0
+        q1 is the second qubit (target)
+        returns a complex numpy array
+        number_of_qubits is the number of qubits in the system.
+        """
+        enlarge_opt = zeros(1 << (num), 1 << (num))  # np.zeros([1 << (num), 1 << (num)])
```

Comment:
I think `2**num` is easier to read than `1<<(num)`

Reply:
skip this one